test/compress/pure_funcs.js

```diff
@@ -304,7 +304,9 @@ issue_2629_1: {
         (/*@__PURE__*/ c)();
         (/*@__PURE__*/ d());
     }
-    expect: {}
+    expect: {
```

Comment:
I think `expect_exact` should be used in this and the tests below with `comments: "all"`.

Reply:
This exposes the known bug where the output of uglify would result in incorrect code if fed to uglify again:
```js
$ echo 'b(); (/*@__PURE__*/ c)(); d();' | bin/uglifyjs -c
b(),c(),d();

$ echo 'b(); (/*@__PURE__*/ c)(); d();' | bin/uglifyjs -c --comments all
b(),/*@__PURE__*/c(),d();

$ echo 'b(); (/*@__PURE__*/ c)(); d();' | bin/uglifyjs -c --comments all | bin/uglifyjs -c
b(),d();
```

Reply:
One workaround could be to emit extra parens around the `AST_Call.expression` if `AST_Call.expression.pure` is set. Another workaround would be to remove all traces of `[@#]__PURE__` from *all* comments when they are output. Still another workaround: as soon as parsed and marked as pure remove the pure comment.

Reply:
Removing comments when not used seems counter-intuitive - and we are still back to the original topic of abusing parentheses to group comments.

IMO this whole feature is a hack much like `pure_funcs`, so I'm not too concerned about the corner cases.

Reply:
> Removing comments when not used seems counter-intuitive - and we are still back to the original topic of abusing parentheses to group comments

Yes, much of this is counter-intuitive. I appreciate that parentheses and comments are a hack, but if `options.output.comments` is not undefined then at output time we could solve all problems by simply replacing `/[@#]__PURE__/g` in all comments with a space character.

Reply:
In any case, if `expect` is used for these tests then the comments and parens should be removed in the output, as they are not accurate.

Reply:
I agree with the use of `expect_exact` - just hit an output bug which I'm working to resolve, that's all.

> if options.output.comments is not undefined then at output time we could solve all problems by simply replacing /[@#]__PURE__/g in all comments with a space character.

We could do that, since it won't present a performance overhead for the default case.